src/distribution.rs

```diff
@@ -0,0 +1,173 @@
+use cosmwasm_std::{coins, Addr, BankMsg, DepsMut, MessageInfo, Response};
+use provwasm_std::ProvenanceQuery;
+
+use crate::{
+    contract::ContractResponse,
+    error::contract_error,
+    msg::Distribution,
+    state::{config_read, outstanding_distributions},
+};
+
+pub fn try_issue_distributions(
+    deps: DepsMut<ProvenanceQuery>,
+    info: MessageInfo,
+    mut distributions: Vec<Distribution>,
+) -> ContractResponse {
+    let state = config_read(deps.storage).load()?;
+
+    if info.sender != state.gp {
+        return contract_error("only gp can issue distributions");
+    }
+
+    if let Some(mut existing) = outstanding_distributions(deps.storage).may_load()? {
+        distributions.append(&mut existing)
+    }
+
+    outstanding_distributions(deps.storage).save(&distributions)?;
+
+    Ok(Response::default())
+}
+
+pub fn try_claim_distribution(
+    deps: DepsMut<ProvenanceQuery>,
+    info: MessageInfo,
+    amount: u64,
+    to: Addr,
+    memo: Option<String>,
+) -> ContractResponse {
+    let state = config_read(deps.storage).load()?;
+
+    let mut distributions = outstanding_distributions(deps.storage).load()?;
+    let distribution = if let Some(index) = distributions
+        .iter()
+        .position(|it| it.subscription == info.sender && it.amount == amount)
+    {
+        distributions.remove(index)
+    } else {
+        return contract_error("no distribution for subscription");
+    };
+
+    outstanding_distributions(deps.storage).save(&distributions)?;
```

Comment:
Is this all in a single "storage" TX? For instance, if the bank send TX after this fails does the distributions collection state get rolled back too?

Reply:
yes rolled back, any writes to storage in general are part of the provenance tx